// @nosto/nosto-react/dist/index.es.js
import { createContext as k, useContext as b, useEffect as g, useRef as h, useMemo as A, cloneElement as x, useState as I, isValidElement as M } from "react";
import { jsx as C } from "react/jsx-runtime";
import { createRoot as _ } from "react-dom/client";
const v = k({
  account: "",
  currentVariation: "",
  responseMode: "HTML",
  clientScriptLoaded: !1
});
function y() {
  return b(v);
}
const P = (e) => String(e) === "[object Object]";
function N(e) {
  if (!P(e)) return !1;
  const t = e.constructor;
  if (t === void 0) return !0;
  const n = t.prototype;
  return !(!P(n) || !n.hasOwnProperty("isPrototypeOf"));
}
function w(e, t) {
  if (e === t)
    return !0;
  if (e instanceof Date && t instanceof Date)
    return e.getTime() === t.getTime();
  if (e instanceof Array && t instanceof Array)
    return e.length !== t.length ? !1 : e.every((n, o) => w(n, t[o]));
  if (N(e) && N(t)) {
    const n = Object.entries(e);
    return n.length !== Object.keys(t).length ? !1 : n.every(([o, r]) => w(r, t[o]));
  }
  return !1;
}
function L(e, t) {
  return g(e, T(t));
}
function T(e) {
  const t = h(e), n = h(0);
  return w(e, t.current) || (t.current = e, n.current += 1), A(() => t.current, [n.current]);
}
function S() {
  window.nostojs = window.nostojs ?? function(e) {
    (window.nostojs.q = window.nostojs.q ?? []).push(e);
  };
}
async function m(e) {
  return window.nostojs(e);
}
typeof window < "u" && (S(), m((e) => {
  e.internal.getSettings();
}));
function D() {
  return typeof window.nosto < "u";
}
const V = {
  production: "https://connect.nosto.com/",
  staging: "https://connect.staging.nosto.com/",
  local: "https://connect.nosto.com/"
};
function O(e) {
  return V[e ?? "production"];
}
function q({ merchantId: e, env: t, options: n, shopifyInternational: o, scriptLoader: r }) {
  const s = document.querySelector("script[nosto-language], script[nosto-market-id]"), c = String(o?.marketId || ""), i = o?.language || "", a = s?.getAttribute("nosto-language") !== i || s?.getAttribute("nosto-market-id") !== c;
  if (!s || a) {
    const u = document.querySelector("#nosto-sandbox");
    s?.parentNode?.removeChild(s), u?.parentNode?.removeChild(u);
    const f = new URL("/script/shopify/market/nosto.js", O(t));
    f.searchParams.append("merchant", e), f.searchParams.append("market", c), f.searchParams.append("locale", i.toLowerCase());
    const E = {
      ...n?.attributes,
      "nosto-language": i,
      "nosto-market-id": c
    };
    return (r ?? j)(f.toString(), { ...n, attributes: E });
  }
  return Promise.resolve();
}
function H(e) {
  if (e.shopifyInternational)
    return q(e);
  const { merchantId: t, env: n, options: o, scriptLoader: r } = e, s = r ?? j, c = new URL(`/include/${t}`, O(n));
  return s(c.toString(), o);
}
function j(e, t) {
  return new Promise((n, o) => {
    const r = document.createElement("script");
    r.src = e, r.async = !0, r.type = "text/javascript", r.onload = () => n(), r.onerror = () => o(), Object.entries(t?.attributes ?? {}).forEach(([s, c]) => r.setAttribute(s, c)), t?.position === "head" ? document.head.appendChild(r) : document.body.appendChild(r);
  });
}
typeof window < "u" && S();
function d(e, t, n) {
  const { clientScriptLoaded: o } = y();
  (n?.deep ? L : g)(() => {
    o && m(e);
  }, [o, ...t ?? []]);
}
function F(e) {
  return x(e.recommendationComponent, {
    // eslint-disable-next-line react/prop-types
    nostoRecommendation: e.nostoRecommendation
  });
}
function R(e) {
  m((t) => t.placements.injectCampaigns(e));
}
function z(e) {
  if (!window.nostojs)
    throw new Error("Nosto has not yet been initialized");
  R(e.recommendations);
}
function l() {
  const { responseMode: e, recommendationComponent: t } = y(), n = h({});
  if (e == "HTML")
    return { renderCampaigns: z };
  if (!t)
    throw new Error("recommendationComponent is required for client-side rendering using hook");
  function o(r) {
    R(r.campaigns?.content ?? {});
    const s = r.campaigns?.recommendations ?? {};
    for (const c in s) {
      const i = s[c], a = "#" + c, u = document.querySelector(a);
      u && (n.current[c] || (n.current[c] = _(u)), n.current[c].render(
        /* @__PURE__ */ C(
          F,
          {
            recommendationComponent: t,
            nostoRecommendation: i
          }
        )
      ));
    }
  }
  return { renderCampaigns: o };
}
function U(e) {
  const { renderCampaigns: t } = l();
  d(async (n) => {
    const o = await n.defaultSession().viewNotFound().setPlacements(e?.placements || n.placements.getPlacements()).load();
    t(o);
  });
}
function ce(e) {
  return U(e), null;
}
function $({ category: e, placements: t }) {
  const { renderCampaigns: n } = l();
  d(
    async (o) => {
      const r = await o.defaultSession().viewCategory(e).setPlacements(t || o.placements.getPlacements()).load();
      n(r);
    },
    [e]
  );
}
function ie(e) {
  return $(e), null;
}
function G(e) {
  const { renderCampaigns: t } = l();
  d(async (n) => {
    const o = await n.defaultSession().viewCart().setPlacements(e?.placements || n.placements.getPlacements()).load();
    t(o);
  });
}
function ae(e) {
  return G(e), null;
}
function J(e) {
  const { renderCampaigns: t } = l();
  d(async (n) => {
    const o = await n.defaultSession().viewFrontPage().setPlacements(e?.placements || n.placements.getPlacements()).load();
    t(o);
  });
}
function ue(e) {
  return J(e), null;
}
function p(e) {
  return !e || typeof e != "object" || W(e) || Z(e) ? e : Array.isArray(e) ? e.map(p) : Object.keys(e).reduce((t, n) => {
    const o = n[0].toLowerCase() + n.slice(1).replace(/([A-Z]+)/g, (r, s) => "_" + s.toLowerCase());
    return t[o] = p(e[n]), t;
  }, {});
}
function W(e) {
  return Object.prototype.toString.call(e) === "[object Date]";
}
function Z(e) {
  return Object.prototype.toString.call(e) === "[object RegExp]";
}
function B({ order: e, placements: t }) {
  const { renderCampaigns: n } = l();
  d(
    async (o) => {
      const r = await o.defaultSession().addOrder(p(e)).setPlacements(t || o.placements.getPlacements()).load();
      n(r);
    },
    [e],
    { deep: !0 }
  );
}
function de(e) {
  return B(e), null;
}
function K(e) {
  const { renderCampaigns: t } = l();
  d(async (n) => {
    const o = await n.defaultSession().viewOther().setPlacements(e?.placements || n.placements.getPlacements()).load();
    t(o);
  });
}
function le(e) {
  return K(e), null;
}
function me({ id: e, pageType: t, children: n }) {
  return /* @__PURE__ */ C("div", { className: "nosto_element", id: e, children: n }, e + (t || ""));
}
function Q({ product: e, tagging: t, placements: n, reference: o }) {
  const { renderCampaigns: r } = l();
  if (t && !t.product_id)
    throw new Error("The product object must contain a product_id property");
  const s = t?.product_id ?? e;
  d(
    async (c) => {
      const i = c.defaultSession().viewProduct(t ?? e).setPlacements(n || c.placements.getPlacements());
      o && i.setRef(s, o);
      const a = await i.load();
      r(a);
    },
    [s, t?.selected_sku_id]
  );
}
function fe(e) {
  return Q(e), null;
}
function X(e, t) {
  return new Promise((n, o) => {
    const r = document.createElement("script");
    r.type = "text/javascript", r.src = e, r.async = !0, r.onload = () => n(), r.onerror = () => o(), Object.entries(t?.attributes ?? {}).forEach(([s, c]) => r.setAttribute(s, c)), t?.position === "head" ? document.head.appendChild(r) : document.body.appendChild(r);
  });
}
const Y = { "nosto-client-script": "" };
function ee(e) {
  const {
    scriptLoader: t = X,
    account: n,
    shopifyMarkets: o,
    loadScript: r = !0
  } = e, [s, c] = I(!1);
  return g(() => {
    function i() {
      c(!0);
    }
    if (S(), m((u) => u.setAutoLoad(!1)), !r) {
      m(i);
      return;
    }
    async function a() {
      await H({
        merchantId: n,
        shopifyInternational: o,
        options: {
          attributes: Y
        },
        scriptLoader: t
      }), i();
    }
    (!D() || o) && a();
  }, [o?.marketId, o?.language]), { clientScriptLoaded: s };
}
function pe(e) {
  const { account: t, multiCurrency: n = !1, children: o, recommendationComponent: r, renderMode: s } = e, c = n ? e.currentVariation : "";
  if (r && !M(r))
    throw new Error(
      "The recommendationComponent prop must be a valid React element. Please provide a valid React element."
    );
  const i = s || (r ? "JSON_ORIGINAL" : "HTML"), { clientScriptLoaded: a } = ee(e);
  return a && m((u) => {
    u.defaultSession().setVariation(c).setResponseMode(i);
  }), /* @__PURE__ */ C(
    v.Provider,
    {
      value: {
        account: t,
        clientScriptLoaded: a,
        currentVariation: c,
        responseMode: i,
        recommendationComponent: r
      },
      children: o
    }
  );
}
function te({ query: e, placements: t }) {
  const { renderCampaigns: n } = l();
  d(
    async (o) => {
      const r = await o.defaultSession().viewSearch(e).setPlacements(t || o.placements.getPlacements()).load();
      n(r);
    },
    [e]
  );
}
function he(e) {
  return te(e), null;
}
function ne({ cart: e, customer: t } = {}) {
  const { clientScriptLoaded: n } = y();
  L(() => {
    const o = e ? p(e) : void 0, r = t ? p(t) : void 0;
    n && m((s) => {
      s.defaultSession().setCart(o).setCustomer(r).viewOther().load({ skipPageViews: !0 });
    });
  }, [n, e, t]);
}
function we(e) {
  return ne(e), null;
}
export {
  ce as Nosto404,
  ie as NostoCategory,
  ae as NostoCheckout,
  v as NostoContext,
  ue as NostoHome,
  de as NostoOrder,
  le as NostoOther,
  me as NostoPlacement,
  fe as NostoProduct,
  pe as NostoProvider,
  he as NostoSearch,
  we as NostoSession,
  U as useNosto404,
  $ as useNostoCategory,
  G as useNostoCheckout,
  y as useNostoContext,
  J as useNostoHome,
  B as useNostoOrder,
  K as useNostoOther,
  Q as useNostoProduct,
  te as useNostoSearch,
  ne as useNostoSession
};
