// @theguild/federation-composition/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyTagFilterOnSubgraphs = exports.addInaccessibleToUnreachableTypes = exports.composeSchemaContract = exports.sortSDL = exports.containsSupergraphSpec = exports.transformSupergraphToPublicSchema = void 0;
__exportStar(require("./compose.js"), exports);
__exportStar(require("./types.js"), exports);
__exportStar(require("./validate.js"), exports);
__exportStar(require("./utils/link/index.js"), exports);
var transform_supergraph_to_public_schema_js_1 = require("./graphql/transform-supergraph-to-public-schema.js");
Object.defineProperty(exports, "transformSupergraphToPublicSchema", { enumerable: true, get: function () { return transform_supergraph_to_public_schema_js_1.transformSupergraphToPublicSchema; } });
var contains_supergraph_spec_js_1 = require("./graphql/contains-supergraph-spec.js");
Object.defineProperty(exports, "containsSupergraphSpec", { enumerable: true, get: function () { return contains_supergraph_spec_js_1.containsSupergraphSpec; } });
var sort_sdl_js_1 = require("./graphql/sort-sdl.js");
Object.defineProperty(exports, "sortSDL", { enumerable: true, get: function () { return sort_sdl_js_1.sortSDL; } });
var schema_contract_js_1 = require("./contracts/schema-contract.js");
Object.defineProperty(exports, "composeSchemaContract", { enumerable: true, get: function () { return schema_contract_js_1.composeSchemaContract; } });
var add_inaccessible_to_unreachable_types_js_1 = require("./contracts/add-inaccessible-to-unreachable-types.js");
Object.defineProperty(exports, "addInaccessibleToUnreachableTypes", { enumerable: true, get: function () { return add_inaccessible_to_unreachable_types_js_1.addInaccessibleToUnreachableTypes; } });
var tag_extraction_js_1 = require("./contracts/tag-extraction.js");
Object.defineProperty(exports, "applyTagFilterOnSubgraphs", { enumerable: true, get: function () { return tag_extraction_js_1.applyTagFilterOnSubgraphs; } });
